mock/mock.go

```diff
@@ -88,15 +96,44 @@ func (c *Call) unlock() {
 	c.Parent.mutex.Unlock()
 }
 
+// If the only return arg is a function which takes and returns Arguments, invoke it instead of returning it as the value
+func (c *Call) getReturnArguments(args Arguments) Arguments {
+	if c.returnFunc != nil {
+		return c.returnFunc(args)
+	}
+
+	return c.ReturnArguments
+}
+
+var argumentsType = reflect.TypeOf(Arguments(nil))
+
 // Return specifies the return arguments for the expectation.
 //
 //    Mock.On("DoSomething").Return(errors.New("failed"))
+//
+// If you pass a single returnArg which is a function that itself takes Arguments and returns Arguments,
+// that will be invoked at runtime.
+//
+//    Mock.On("HelloWorld").Return(func(args mock.Arguments) mock.Arguments { return "Hello " + arg[0].(string) })
 func (c *Call) Return(returnArguments ...interface{}) *Call {
 	c.lock()
 	defer c.unlock()
 
 	c.ReturnArguments = returnArguments
 
+	if len(c.ReturnArguments) == 1 {
```

Comment:
I love this PR, but you're doing a lot of unnecessary work here.  A type assertion will get the job done:

```go
func (c *Call) Return(returnArguments ...interface{}) *Call {
	c.lock()
	defer c.unlock()

	if len(returnArguments) == 1 {
		if fn, ok := returnArguments[0].(func(Arguments) Arguments); ok {
			c.returnFunc = fn
			return c
		}
	}

	c.returnFunc = nil
	c.ReturnArguments = returnArguments
	return c
}
```

Reply:
gotcha I think I'd tried to typecast to my returnArgumentsFunc type which doesn't work unless the caller explicitly uses that type - but casting against the anonymous `func(Arguments) Arguments` indeed works